### DISCUSSION TITLE
Use a generic typevar in Mapped column

### DISCUSSION THREAD

**Discussion by Alc-Alc** [sqlalchemy/sqlalchemy#9700] [Usage Questions] [answered]
I'd like to have a generic column `T` that can be assigned different types in subclasses. The expected output would be that table `my_int_impl` has a column `my_data` with `Integer` type and `my_str_impl` has a column `my_data` with `String` type.

Right now this raises the following error

```
AttributeError: 'TypeVar' object has no attribute '__mro__'. Did you mean: '__or__'?
```

Code

```python
from typing import TypeVar, Generic
from sqlalchemy.orm import Mapped, mapped_column, MappedAsDataclass, DeclarativeBase
from sqlalchemy import create_engine

T = TypeVar("T")

class MyGenericMixin(Generic[T]):
    id: Mapped[int] = mapped_column(primary_key=True)
    my_data: Mapped[T]

class Base(MappedAsDataclass, DeclarativeBase):
    pass

class MyIntImpl(Base, MyGenericMixin[int]):
    __tablename__ = "my_int_impl"

class MyStrImpl(Base, MyGenericMixin[str]):
    __tablename__ = "my_str_impl"

engine = create_engine("sqlite://")
Base.metadata.create_all(engine)
```

Disclaimer

This question is ~~copied~~ inspired from this [question on Stack Overflow](https://stackoverflow.com/questions/76080957/sqlalchemy-2-generic-mixin). While the OP there has posted a solution, I'd like to know the standard approach.

**Comment by zzzeek** [comment, marked as answer]
There's two answers here, one is if you want this to be **fully automatic at runtime**, which is not really possible exactly as written, there may be ways to use special mixins that can figure this out to some extent, or you just want **the typing to work out this way**, which is always possible by using ``TYPE_CHECKING``.

So answer one, "fully automatic at runtime", SQLAlchemy does not implement this feature right now, and without some custom mixins being used, the specific generic variable "int", "str", etc. is not available at runtime from the typing.Generic object:

```py
>>> from typing import TypeVar, Generic
>>> T = TypeVar("T")
>>> class Foo(Generic[T]):
...     pass
... 
>>> Foo.__mro__

# note "_T" is not here, this is the Generic class by itself
(<class '__main__.Foo'>, <class 'typing.Generic'>, <class 'object'>)
>>> Foo.__mro__[1] is Generic
True

```

Beyond that, as well as whether or not the ORM fails more gracefully when it sees "_T" (that should be fixed), there's no intrinsic way at runtime, short of reinventing all of mypy within runtime interpretation, that could figure out that this T links to the particular type variable that's inside of MyGenericMixin in the upper class' MRO.  Again, some special mixin that "knows" the indicated sub-type and "knows" to look through all attributes and set them up as needed could maybe do this.
   
Such a use case that works this way, that is, fully automatically, would need to be implemented at the runtime class level for "MyGenericMixin", likely using the `__init_subclass__()` hook to replace the `my_data` annotation in the class before the class is delivered to Declarative.   This code wont work but this is the general form:

```py
class PutAConcreteTypeIn(...):
    def __init_subclass__(cls, **kw):
        figure_out_what_type_to_use()
        cls.my_data = mapped_column(....)
        super().__init_subclass__(**kw)


class MyGenericMixin(PutAConcreteTypeIn[T]):
    id: Mapped[int] = mapped_column(primary_key=True)
    my_data: Mapped[T]


class Base(DeclarativeBase):
    pass


class MyIntImpl(MyGenericMixin[int], Base):
    __tablename__ = "my_int_impl"
```

note the mixin class needs to occur before Base in order for the `__init_subclass__()` hook to be called at the right time.

To get closer towards this being possible, we probably would need to use a class other than Generic in order to get the "Generic"-ness, which maintains the inner type at runtime.   Would have to play with some class mechanics to figure that out.



To do this pattern just at the typing level only, use TYPE_CHECKING:


```py
T = TypeVar("T")

class MyGenericMixin(Generic[T]):
    id: Mapped[int] = mapped_column(primary_key=True)

    if TYPE_CHECKING:
       my_data: Mapped[T]

class Base(DeclarativeBase):
    pass

class MyIntImpl(Base, MyGenericMixin[int]):
    __tablename__ = "my_int_impl"

    my_data: Mapped[int]

class MyStrImpl(Base, MyGenericMixin[str]):
    __tablename__ = "my_str_impl"

    my_data: Mapped[str]
```

you could even do this:

```py
class MyIntImpl(Base, MyGenericMixin[int]):
    __tablename__ = "my_int_impl"

    if not TYPE_CHECKING:
        my_data: Mapped[int]
```


There might be some other more creative ways for SQLAlchemy ORM to interpret "_T" here, right now it's failing and it should at least anticipate this variable and raise an informative error .  if a special mixin could be devised which maintains the class-level type variable, the informative error could refer to this extension.

**Reply by Alc-Alc** [reply]
Thanks for the detailed explanation 😇

**Comment by CaselIT** [comment]
Hi,

I'm not sure if it's doable, since I don't think python replaces the typevar with the actual type at runtime, so I believe the typevar is what's always seen at runtime:

```py
from typing import Generic, TypeVar, get_type_hints

T = TypeVar("T")

class Base(Generic[T]):
    x: T

class Sub(Base[int]):
    y: str

print(Base.__annotations__, get_type_hints(Base))
print(Sub.__annotations__, get_type_hints(Sub))
```

This prints the following:
```
{'x': ~T} {'x': ~T}
{'y': <class 'str'>} {'x': ~T, 'y': <class 'str'>}
```
